Added functions to undraw all specific entities

Feuerbowle opened the pull request:
New functions to undraw all:
- Toolpaths
- Tools
- Boundaries
- Patterns
- Feature Groups
- Feature Sets
- Workplanes
- Levels and Sets
- Stockmodels

also added test classes for:
- Feature Groups
- Feature Sets
- Workplanes
- Levels and Sets
- Stockmodels

Feuerbowle committed:
Added functions to undraw all specific entities

New functions to undraw all:

- Toolpaths
- Tools
- Boundaries
- Patterns
- Feature Groups
- Feature Sets
- Workplanes
- Levels and Sets
- Stockmodels

also added test classes for:
- Feature Groups
- Feature Sets
- Workplanes
- Levels and Sets
- Stockmodels

lukeedw commented on Delcam.ProductInterface.PowerMILL.Test/PMEntityTests/PMBoundaryTests.cs:
```diff
@@ -208,6 +208,15 @@ public void InsertToolpathTest()
+        public void UndrawAllBoundariesTest()
+        {
+            _powerMILL.LoadProject(TestFiles.BoundaryTypes);
+            _powerMILL.Execute("DRAW BOUNDARY ALL");
```
Perhaps add a function for DrawAll while you are doing UndrawAll?

lukeedw commented on Delcam.ProductInterface.PowerMILL.Test/PMEntityTests/PMFeatureGroupTests.cs:
```diff
@@ -0,0 +1,68 @@
+        [Test]
+        public void UndrawAllFeatureGroupsTest()
+        {
+            _powerMill.Execute("CREATE FEATUREGROUP ;");
```
Can you create these via the API or do we have to use Macro calls?

lukeedw commented on Delcam.ProductInterface.PowerMILL.Test/PMEntityTests/PMFeatureSetTests.cs:
```diff
@@ -0,0 +1,68 @@
+        [Test]
+        public void UndrawAllFeatureSetsTest()
+        {
+            _powerMill.Execute("CREATE FEATURESET ;");
```
Can you create these via the API or do we have to use Macro calls?

lukeedw commented on Delcam.ProductInterface.PowerMILL.Test/PMEntityTests/PMLevelOrSetTests.cs:
```diff
@@ -0,0 +1,68 @@
+        [Test]
+        public void UndrawAllLevelsOrSetsTest()
+        {
+            _powerMill.Execute("CREATE LEVEL ;");
```
I'm sure these could be created via the API.

lukeedw commented on Delcam.ProductInterface.PowerMILL.Test/PMEntityTests/PMStockmodelTests.cs:
```diff
@@ -0,0 +1,68 @@
+        [Test]
+        public void UndrawAllStockmodelsTest()
+        {
+            _powerMill.Execute("CREATE STOCKMODEL ;");
```
Can this be created via API.

lukeedw commented on Delcam.ProductInterface.PowerMILL.Test/PMEntityTests/PMWorkplaneTests.cs:
```diff
@@ -0,0 +1,68 @@
+        [Test]
+        public void UndrawAllWorkplaneTest()
+        {
+            _powerMill.Execute("CREATE WORKPLANE ;");
```
These could be created via the API.

lukeedw commented on Delcam.ProductInterface.PowerMILL/Classes/PMBoundariesCollection.cs:
```diff
@@ -184,6 +184,14 @@ public PMBoundary CreateBoundary(Spline spline)
+        /// </summary>        
+        public void UndrawAll()
+        {
+            _powerMILL.DoCommand("UNDRAW BOUNDARY ALL");
```
I think you could simplify all of these by putting the function on the base class PMEntity and use the code:
`_powerMILL.DoCommand("UNDRAW` " + Identifier + " ALL");

Feuerbowle replied:
Sorry for my long abstinence, but there was too much regular work to do.

You're right, that's an much cleaner and easier way to do this. But if i'm adding the function to the PMEntity class, the function would only be accessible from an specific entity.

e.g.
```
PMToolpath TPTest;   
TPTest.Undraw;
```

This is fine and i will add this function later, but at the moment i'm at the Undraw**ALL** function. So i think i must insert that function in the PMEntitiesCollection class. But here i encountered one problem. There is no "Idenifier" at the PMEntitiesCollection class. The only "workaround" i've found is this:

```
public void UndrawAll()
 {             
     if (Count > 0)
     {                
         _powerMILL.DoCommand("UNDRAW " + this[0].Identifier + " ALL");
     }
 }
```
Is there another, easier way to get the identifier?

lukeedw replied:
I believe you should be able to use:
`T.Identifier`
It is a generic class and T gives you access to the actual type at runtime.

Feuerbowle replied:
I tried it but i'm getting this error:
```
CS0119: 'T' is a 'type parameter', which is not valid in this context"
```

lukeedw replied:
Hmmm, apparently that's not allowed:
https://stackoverflow.com/questions/196661/calling-a-static-method-on-a-generic-type-parameter
There is some discussion in the above post that you could either use reflection or else use:
`new T().Identifer`

Feuerbowle replied:
OK, this is uncharted territories for me. Will try to figure this out. Thanks for the hint

lukeedw reviewed (commented):
I think you can simplify the code to undraw as well as adding a DrawAll function.  Also see the comments about how you are creating the test entities.

Feuerbowle committed:
Changed DrawAll & UndrawAll function

Changed the DrawAll and UndrawAll Function to the PMEntitiesCollection class.

The function now uses reflection to determine the entity Indetifier

lukeedw commented on Delcam.ProductInterface.PowerMILL/Classes/PMEntitiesCollection.cs:
```diff
@@ -174,6 +174,28 @@ public string GetNewEntityName(string basePrefix = null)
+        public void DrawAll()
+        {
+            System.Reflection.FieldInfo[] fi = typeof(T).GetFields(System.Reflection.BindingFlags.NonPublic | System.Reflection.BindingFlags.Static);
+            string Identifier = fi[0].GetValue(null).ToString();
```
Small thing, can you change Identifier to identifier.  Upper case first character implies it is a Property rather than a local variable.

Feuerbowle replied:
Sure, i will do that.

lukeedw commented on Delcam.ProductInterface.PowerMILL/Classes/PMEntitiesCollection.cs:
```diff
@@ -174,6 +174,28 @@ public string GetNewEntityName(string basePrefix = null)
+        {
+            System.Reflection.FieldInfo[] fi = typeof(T).GetFields(System.Reflection.BindingFlags.NonPublic | System.Reflection.BindingFlags.Static);
+            string Identifier = fi[0].GetValue(null).ToString();
+            if (Identifier != "MODEL")
```
What happens if it is a Model?

Feuerbowle replied:
OK, i've done some testings:

**UNDRAW MODEL ALL**
This command works and seems to be the same as
```
VIEW MODEL ; SHADE OFF
VIEW MODEL ; WIREFRAME OFF
```
So we can delete the if condition in the Undraw Function


**DRAW MODEL ALL**
This will cause an Error:
```
Command Error:
DRAW MODEL ALL
Unsupported feature or command
```

lukeedw replied:
I think the command for models might just be **DRAW ALL**

Feuerbowle replied:
**DRAW ALL** will draw all entities

We could use

```
VIEW MODEL ; SHADE ON
VIEW MODEL ; WIREFRAME ON
```

Feuerbowle committed:
Minor changes as requested

- Changed "Identifier" to "identifier"
- Added functionality to handle Draw and Undraw Models
- Added function to create FeatureSets and FeatureGroups with API
- Updated Test routines
